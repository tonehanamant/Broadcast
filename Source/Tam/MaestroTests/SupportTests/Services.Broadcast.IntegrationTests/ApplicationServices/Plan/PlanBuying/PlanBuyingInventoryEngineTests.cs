using ApprovalTests;
using ApprovalTests.Reporters;
using Newtonsoft.Json;
using NUnit.Framework;
using Services.Broadcast.BusinessEngines;
using Services.Broadcast.Clients;
using Services.Broadcast.Entities;
using Services.Broadcast.Entities.Enums;
using Services.Broadcast.Entities.Plan.Buying;
using Services.Broadcast.IntegrationTests.Helpers;
using Services.Broadcast.IntegrationTests.Stubs;
using Services.Broadcast.Repositories;
using System.Collections.Generic;
using System.Linq;
using Tam.Maestro.Common.DataLayer;
using Tam.Maestro.Data.Entities.DataTransferObjects;
using Unity;
using static Services.Broadcast.Entities.Plan.Buying.PlanBuyingInventoryProgram;
using static Services.Broadcast.Entities.Plan.CommonPricingEntities.BasePlanInventoryProgram;
using static Services.Broadcast.Entities.Plan.PlanDaypartDto;

namespace Services.Broadcast.IntegrationTests.ApplicationServices.Plan.PlanBuying
{
    [TestFixture]
    [UseReporter(typeof(DiffReporter))]
    public class PlanBuyingInventoryEngineTests
    {
        private readonly IPlanBuyingInventoryEngine _PlanBuyingInventoryEngine;
        private readonly IPlanRepository _PlanRepository;
        private readonly IStationRepository _StationRepository;
        private readonly InventoryFileTestHelper _InventoryFileTestHelper;

        public PlanBuyingInventoryEngineTests()
        {
            _PlanBuyingInventoryEngine = IntegrationTestApplicationServiceFactory.GetApplicationService<IPlanBuyingInventoryEngine>();
            _PlanRepository = IntegrationTestApplicationServiceFactory.BroadcastDataRepositoryFactory.GetDataRepository<IPlanRepository>();
            _StationRepository = IntegrationTestApplicationServiceFactory.BroadcastDataRepositoryFactory.GetDataRepository<IStationRepository>();
            _InventoryFileTestHelper = new InventoryFileTestHelper();

            var launchDarklyClientStub = (LaunchDarklyClientStub)IntegrationTestApplicationServiceFactory.Instance.Resolve<ILaunchDarklyClient>();
            // TODO SDE : this should be reworked for these to be true, as they are in production
            //launchDarklyClientStub.FeatureToggles[FeatureToggles.ALLOW_MULTIPLE_CREATIVE_LENGTHS] = false;
        }

        [Test]
        [Category("short_running")]
        public void GetsFallbackInventoryForBuying()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1196);
                plan.CreativeLengths.First().SpotLengthId = 2;
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan, 
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [Category("long_running")]
        public void GetsInventoryForBuying()
        {
            try
            {
                StubbedConfigurationWebApiClient.RunTimeParameters["PlanBuyingEndpointVersion"] = "3";

                using (new TransactionScopeWrapper())
                {
                    var diagnostic = new PlanBuyingJobDiagnostic();
                    var plan = _PlanRepository.GetPlan(1198);

                    plan.CreativeLengths = new List<CreativeLength>
                    {
                        new CreativeLength
                        {
                            SpotLengthId = 1,
                            Weight = 50
                        },
                        new CreativeLength
                        {
                            SpotLengthId = 2,
                            Weight = 50
                        }
                    };

                    var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                        plan,
                        new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                        _GetAvailableInventorySources(),
                        diagnostic);

                    var jsonResolver = new IgnorableSerializerContractResolver();
                    jsonResolver.Ignore(typeof(PlanBuyingInventoryProgram), "ManifestId");
                    jsonResolver.Ignore(typeof(ManifestDaypart), "Id");
                    jsonResolver.Ignore(typeof(ManifestWeek), "Id");
                    var jsonSettings = new JsonSerializerSettings()
                    {
                        ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                        ContractResolver = jsonResolver
                    };
                    var json = IntegrationTestHelper.ConvertToJson(result, jsonSettings);

                    Approvals.Verify(json);
                }
            }
            finally
            {
                StubbedConfigurationWebApiClient.RunTimeParameters["PlanBuyingEndpointVersion"] = "2";
            }
        }

        [Test]
        [Category("long_running")]
        public void RunWithHiatusDayTest()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1199);
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan, 
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [Category("long_running")]
        public void RunSpotLengthTest()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1197);
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan, 
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [Category("long_running")]
        public void RunAffilitateContentRestrictionTest()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1197);
                plan.Dayparts[0].Restrictions.AffiliateRestrictions = new RestrictionsDto.AffiliateRestrictionsDto
                {
                    Affiliates = new List<LookupDto>()
                };
                plan.Dayparts[0].Restrictions.AffiliateRestrictions.Affiliates.Add(new LookupDto(8, "CW"));
                plan.Dayparts[0].Restrictions.AffiliateRestrictions.ContainType = ContainTypeEnum.Exclude;
                _PlanRepository.SavePlan(plan, "IntegrationTestUser", new System.DateTime(2020, 11, 3));
                // Result has affiliates CW and FOX.
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan,
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [Category("long_running")]
        public void RunAffilitateContentRestrictionUpdateMonthDetailsTest()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1197);
                _StationRepository.SaveStationMonthDetails(new Entities.StationMonthDetailDto
                {
                    Affiliation = "CW2",
                    MediaMonthId = 500,
                    DistributorCode = 39,
                    MarketCode = 101,
                    StationId = 575
                });
                // Since affiliate has been updated, it shouldn't be filtered by inventory gathering.
                plan.Dayparts[0].Restrictions.AffiliateRestrictions = new RestrictionsDto.AffiliateRestrictionsDto
                {
                    Affiliates = new List<LookupDto>()
                };
                plan.Dayparts[0].Restrictions.AffiliateRestrictions.Affiliates.Add(new LookupDto(8, "CW"));
                plan.Dayparts[0].Restrictions.AffiliateRestrictions.ContainType = ContainTypeEnum.Exclude;
                _PlanRepository.SavePlan(plan, "IntegrationTestUser", new System.DateTime(2020, 11, 3));
                // Result has affiliates CW and FOX.
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan,
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [Category("long_running")]
        public void GetInventoryForPlanFlightDaysTest()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1197);
                plan.FlightDays.Clear();
                plan.FlightDays.Add(1);
                plan.FlightDays.Add(5);
                _PlanRepository.SavePlan(plan, "IntegrationTestUser", new System.DateTime(2020, 2, 27));
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan,
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        [Test]
        [Category("long_running")]
        public void RunWithProgramWithMultipleAudiencesTest()
        {
            using (new TransactionScopeWrapper())
            {
                var diagnostic = new PlanBuyingJobDiagnostic();
                var plan = _PlanRepository.GetPlan(1200);
                var result = _PlanBuyingInventoryEngine.GetInventoryForPlan(
                    plan, 
                    new PlanBuyingInventoryEngine.ProgramInventoryOptionalParametersDto(),
                    _GetAvailableInventorySources(),
                    diagnostic);

                Approvals.Verify(IntegrationTestHelper.ConvertToJson(result));
            }
        }

        private List<int> _GetAvailableInventorySources()
        {
            return new List<int>
            {
                1,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12
            };
        }
    }
}
