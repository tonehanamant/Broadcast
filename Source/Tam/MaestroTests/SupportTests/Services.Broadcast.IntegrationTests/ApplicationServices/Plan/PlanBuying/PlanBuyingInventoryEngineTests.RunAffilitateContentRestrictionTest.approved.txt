[
  {
    "Cpm": 119.04761904761904761904761900,
    "Impressions": 210.0,
    "InventoryBuyingQuarterType": "Plan",
    "InventorySource": {
      "Id": 1,
      "InventoryType": 1,
      "IsActive": true,
      "Name": "Open Market"
    },
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 300.0
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0
      },
      {
        "AudienceId": 37,
        "Impressions": 100.0
      },
      {
        "AudienceId": 39,
        "Impressions": 100.0
      },
      {
        "AudienceId": 40,
        "Impressions": 300.0
      }
    ],
    "ManifestDayparts": [
      {
        "Daypart": {
          "_Code": "CUS",
          "_EndTime": 1799,
          "_Friday": true,
          "_Id": 32941,
          "_Is24Hour": false,
          "_LastEndTime": 0,
          "_LastStartTime": 0,
          "_Monday": true,
          "_Name": "BRDCST DP362",
          "_Saturday": false,
          "_StartTime": 0,
          "_Sunday": false,
          "_Thursday": true,
          "_Tuesday": true,
          "_Wednesday": true
        },
        "Id": 360318,
        "PrimaryProgram": {
          "EndTime": 7200,
          "Genre": "Comedy",
          "Id": 11250,
          "Name": "Late news",
          "ShowType": "Sports",
          "StartTime": 3600
        },
        "PrimaryProgramId": 11250,
        "ProgramName": "TWO & HALF MEN",
        "Programs": [
          {
            "EndTime": 7200,
            "Genre": "Comedy",
            "Id": 11250,
            "Name": "Late news",
            "ShowType": "Sports",
            "StartTime": 3600
          }
        ]
      }
    ],
    "ManifestId": 288692,
    "ManifestRates": [
      {
        "Cost": 50.0000,
        "SpotLengthId": 2
      }
    ],
    "ManifestWeeks": [
      {
        "ContractMediaWeekId": 771,
        "Id": 1698584,
        "InventoryMediaWeekId": 771,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 772,
        "Id": 1705866,
        "InventoryMediaWeekId": 772,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 773,
        "Id": 1713195,
        "InventoryMediaWeekId": 773,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 774,
        "Id": 1720549,
        "InventoryMediaWeekId": 774,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 775,
        "Id": 1728281,
        "InventoryMediaWeekId": 775,
        "Spots": 0
      }
    ],
    "NsiToNtiImpressionConversionRate": 0.7,
    "PostingType": 1,
    "ProjectedImpressions": 61.39,
    "ProvidedImpressions": 210.0,
    "SpotLengthId": 1,
    "StandardDaypartId": 2,
    "Station": {
      "Affiliation": "FOX",
      "CallLetters": "WPFO 23.1",
      "Code": 8346,
      "Id": 2106,
      "IsTrueInd": false,
      "LegacyCallLetters": "WPFO",
      "ManifestMaxEndDate": null,
      "MarketCode": 100,
      "ModifiedDate": null,
      "OriginMarket": null,
      "OwnershipGroupName": "Ownership group 3",
      "RateDataThrough": null,
      "RepFirmName": "Sales group 4"
    },
    "Unit": null
  },
  {
    "Cpm": 28.571428571428571428571428600,
    "Impressions": 210.0,
    "InventoryBuyingQuarterType": "Plan",
    "InventorySource": {
      "Id": 1,
      "InventoryType": 1,
      "IsActive": true,
      "Name": "Open Market"
    },
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 300.0
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0
      },
      {
        "AudienceId": 37,
        "Impressions": 300.0
      },
      {
        "AudienceId": 39,
        "Impressions": 300.0
      },
      {
        "AudienceId": 40,
        "Impressions": 400.0
      }
    ],
    "ManifestDayparts": [
      {
        "Daypart": {
          "_Code": "CUS",
          "_EndTime": 3599,
          "_Friday": false,
          "_Id": 29555,
          "_Is24Hour": false,
          "_LastEndTime": 0,
          "_LastStartTime": 0,
          "_Monday": false,
          "_Name": "BRDCST DP362",
          "_Saturday": true,
          "_StartTime": 0,
          "_Sunday": false,
          "_Thursday": false,
          "_Tuesday": false,
          "_Wednesday": false
        },
        "Id": 360319,
        "PrimaryProgram": {
          "EndTime": 7200,
          "Genre": "Comedy",
          "Id": 11251,
          "Name": "Late news",
          "ShowType": "Sports",
          "StartTime": 3600
        },
        "PrimaryProgramId": 11251,
        "ProgramName": "2.5 MEN X2",
        "Programs": [
          {
            "EndTime": 7200,
            "Genre": "Comedy",
            "Id": 11251,
            "Name": "Late news",
            "ShowType": "Sports",
            "StartTime": 3600
          }
        ]
      }
    ],
    "ManifestId": 288693,
    "ManifestRates": [
      {
        "Cost": 12.0000,
        "SpotLengthId": 2
      }
    ],
    "ManifestWeeks": [
      {
        "ContractMediaWeekId": 771,
        "Id": 1698585,
        "InventoryMediaWeekId": 771,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 772,
        "Id": 1705867,
        "InventoryMediaWeekId": 772,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 773,
        "Id": 1713196,
        "InventoryMediaWeekId": 773,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 774,
        "Id": 1720550,
        "InventoryMediaWeekId": 774,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 775,
        "Id": 1728282,
        "InventoryMediaWeekId": 775,
        "Spots": 0
      }
    ],
    "NsiToNtiImpressionConversionRate": 0.7,
    "PostingType": 1,
    "ProjectedImpressions": 152.11875,
    "ProvidedImpressions": 210.0,
    "SpotLengthId": 1,
    "StandardDaypartId": 2,
    "Station": {
      "Affiliation": "FOX",
      "CallLetters": "WPFO 23.1",
      "Code": 8346,
      "Id": 2106,
      "IsTrueInd": false,
      "LegacyCallLetters": "WPFO",
      "ManifestMaxEndDate": null,
      "MarketCode": 100,
      "ModifiedDate": null,
      "OriginMarket": null,
      "OwnershipGroupName": "Ownership group 3",
      "RateDataThrough": null,
      "RepFirmName": "Sales group 4"
    },
    "Unit": null
  },
  {
    "Cpm": 42.857142857142857142857142900,
    "Impressions": 140.0,
    "InventoryBuyingQuarterType": "Plan",
    "InventorySource": {
      "Id": 1,
      "InventoryType": 1,
      "IsActive": true,
      "Name": "Open Market"
    },
    "ManifestAudiences": [
      {
        "AudienceId": 31,
        "Impressions": 200.0
      },
      {
        "AudienceId": 33,
        "Impressions": 100.0
      },
      {
        "AudienceId": 37,
        "Impressions": 100.0
      },
      {
        "AudienceId": 39,
        "Impressions": 100.0
      },
      {
        "AudienceId": 40,
        "Impressions": 200.0
      }
    ],
    "ManifestDayparts": [
      {
        "Daypart": {
          "_Code": "CUS",
          "_EndTime": 3599,
          "_Friday": false,
          "_Id": 29548,
          "_Is24Hour": false,
          "_LastEndTime": 0,
          "_LastStartTime": 0,
          "_Monday": false,
          "_Name": "BRDCST DP362",
          "_Saturday": false,
          "_StartTime": 0,
          "_Sunday": true,
          "_Thursday": false,
          "_Tuesday": false,
          "_Wednesday": false
        },
        "Id": 360320,
        "PrimaryProgram": {
          "EndTime": 7200,
          "Genre": "Comedy",
          "Id": 11252,
          "Name": "Late news",
          "ShowType": "Sports",
          "StartTime": 3600
        },
        "PrimaryProgramId": 11252,
        "ProgramName": "SEINFELD X2",
        "Programs": [
          {
            "EndTime": 7200,
            "Genre": "Comedy",
            "Id": 11252,
            "Name": "Late news",
            "ShowType": "Sports",
            "StartTime": 3600
          }
        ]
      }
    ],
    "ManifestId": 288694,
    "ManifestRates": [
      {
        "Cost": 12.0000,
        "SpotLengthId": 2
      }
    ],
    "ManifestWeeks": [
      {
        "ContractMediaWeekId": 771,
        "Id": 1698586,
        "InventoryMediaWeekId": 771,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 772,
        "Id": 1705868,
        "InventoryMediaWeekId": 772,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 773,
        "Id": 1713197,
        "InventoryMediaWeekId": 773,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 774,
        "Id": 1720551,
        "InventoryMediaWeekId": 774,
        "Spots": 0
      },
      {
        "ContractMediaWeekId": 775,
        "Id": 1728283,
        "InventoryMediaWeekId": 775,
        "Spots": 0
      }
    ],
    "NsiToNtiImpressionConversionRate": 0.7,
    "PostingType": 1,
    "ProjectedImpressions": 203.0,
    "ProvidedImpressions": 140.0,
    "SpotLengthId": 1,
    "StandardDaypartId": 2,
    "Station": {
      "Affiliation": "FOX",
      "CallLetters": "WPFO 23.1",
      "Code": 8346,
      "Id": 2106,
      "IsTrueInd": false,
      "LegacyCallLetters": "WPFO",
      "ManifestMaxEndDate": null,
      "MarketCode": 100,
      "ModifiedDate": null,
      "OriginMarket": null,
      "OwnershipGroupName": "Ownership group 3",
      "RateDataThrough": null,
      "RepFirmName": "Sales group 4"
    },
    "Unit": null
  }
]